src/integer_mod_q/mat_zq/inverse.rs

```diff
@@ -0,0 +1,194 @@
+    ///
+    /// # Panics ...
+    /// - if the number of rows is not equal to the number of columns.
+    /// - if the inverse of an entry can not be computed because the modulus is not prime.
+    pub fn inverse_prime(&self) -> Option<MatZq> {
+        // Check if determinant is coprime to modulus.
+        let det = MatZ::from(self).det().ok()?;
+        if det.gcd(self.get_mod()) != Z::ONE {
+            None
+        } else {
+            let dimensions = self.get_num_rows();
+
+            // Concatenate the matrix with the identity matrix.
+            let matrix_identity = self
+                .concat_horizontal(&MatZq::identity(dimensions, dimensions, self.get_mod()))
+                .unwrap();
+
+            let identity_inverse = matrix_identity.gaussian_elimination_prime();
+
+            // The inverse is now the right half of the matrix `identity_inverse`.
+            let mut inverse = MatZq::new(dimensions, dimensions, self.get_mod());
+            for i in 0..dimensions {
+                inverse
+                    .set_column(i, &identity_inverse, dimensions + i)
+                    .unwrap();
+            }
+            Some(inverse)
+        }
+    }
+
+    /// Returns the `row echelon form` of the matrix using gaussian elimination or
+    /// panics if the modulus is not prime.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let mut matrix = MatZq::from_str("[[1,2],[3,4]] mod 7").unwrap();
+    /// let matrix_gauss = matrix.gaussian_elimination_prime();
+    ///
+    /// assert_eq!("[[1, 0],[0, 1]] mod 7", matrix_gauss.to_string());
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if the modulus is not prime.
+    pub fn gaussian_elimination_prime(self) -> MatZq {
+        if !self.get_mod().is_prime() {
+            panic!("The modulus of the matrix is not prime");
+        }
+
+        // Since we only want the echelon form, the permutation `perm` is not relevant.
+        let mut perm: i64 = 1;
+        unsafe { fmpz_mod_mat_rref(&mut perm, &self.matrix) };
+
+        self
+    }
+}
+
+#[cfg(test)]
+mod test_inverse {
+    use crate::integer_mod_q::MatZq;
+    use std::str::FromStr;
+
+    /// Test whether `inverse` correctly calculates an inverse matrix.
+    #[test]
+    fn inverse_works() {
+        let mat = MatZq::from_str("[[5,2],[2,1]] mod 7").unwrap();
+
+        let inv = mat.inverse_prime().unwrap();
+
+        let cmp_inv = MatZq::from_str("[[1, 5],[5, 5]] mod 7").unwrap();
+        assert_eq!(cmp_inv, inv);
+    }
+
+    /// Check if the multiplication of inverse and matrix result in an identity matrix.
+    #[test]
+    fn inverse_correct() {
+        let mat = MatZq::from_str("[[5,2],[2,1]] mod 11").unwrap();
+
+        let inv = mat.inverse_prime().unwrap();
+        let diag = mat * inv;
+
+        let cmp = MatZq::from_str("[[1,0],[0,1]] mod 11").unwrap();
+        assert_eq!(cmp, diag);
+    }
```

Comment:
Why are `inverse_works` and `inverse_correct` two different test cases? They test the same matrix.

Reply:
The first test is for the correct inverse, the second one assures that the inverse multiplied by the matrix results in an identity matrix